### PR TITLE
chore: Convert to controller-runtime's new `reconcile.AsReconciler`

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #N/A 

**Description**

Removes our `controller.TypedController` and replaces it with controller-runtime's typed controller implementation e.g. [`reconcile.AsReconciler()`](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.2/pkg/reconcile/reconcile.go#L123)

Takes in the upstream change from https://github.com/kubernetes-sigs/karpenter/pull/1247

**How was this change tested?**

`make presubmit`
`/karpenter snapshot`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

File: pkg/controllers/nodeclass/status/securitygroup_test.go [hunk 1]
```diff
@@ -219,10 +218,7 @@ var _ = Describe("NodeClass Security Group Status Controller", func() {
 			},
 		}
 		ExpectApplied(ctx, env.Client, nodeClass)
-		ExpectReconcileSucceeded(ctx, statusController, client.ObjectKeyFromObject(nodeClass))
+		ExpectObjectReconciled(ctx, env.Client, statusController, nodeClass)
 		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
-		Expect(nodeClass.Status.SecurityGroups).To(BeNil())
-		Expect(nodeClass.StatusConditions().Get(v1beta1.ConditionTypeNodeClassReady).IsFalse()).To(BeTrue())
-		Expect(nodeClass.StatusConditions().Get(v1beta1.ConditionTypeNodeClassReady).Message).To(Equal("unable to resolve security groups"))
```
Comment (jmdeal) [inline pkg/controllers/nodeclass/status/securitygroup_test.go:226]: Why drop these expectations?

File: test/pkg/debug/monitor.go [hunk 2]
```diff
@@ -80,8 +79,12 @@ func (m *Monitor) Stop() {
 	m.wg.Wait()
 }
 
-func newControllers(kubeClient client.Client) []controller.Controller {
-	return []controller.Controller{
+func newControllers(kubeClient client.Client) []interface {
```
Comment (jmdeal) [inline test/pkg/debug/monitor.go:82]: nit: replace inline interface with upstream `controller.Controller`

File: pkg/controllers/nodeclass/status/securitygroup_test.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclass/status/securitygroup_test.go:226]: Probably a bad rebase.

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

**Review by jmdeal — APPROVED** [review]
LGTM 🚀

**Pull request merged**